### PR TITLE
4755 – Add otel collector

### PR DESCRIPTION
**Pull request by vasconsaurus**
## Description

This adds `otel-collector`. It scrapes the Prometheus endpoint for our metrics and sends it to Honeycomb.

I added the bare minimum configuration for it to work. There are features that might be interesting, but I think we could add those as we add the metrics and understand what makes sense for us.

- Adding a batch processor might be interesting: https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/batchprocessor/README.md
- Maybe activating internal telemetry? https://opentelemetry.io/docs/collector/internal-telemetry#activate-internal-telemetry-in-the-collector

**Resources**
- https://docs.honeycomb.io/integrations/metrics/prometheus/
- https://docs.honeycomb.io/send-data/opentelemetry/collector/
- https://github.com/open-telemetry/opentelemetry-collector
- https://opentelemetry.io/docs/collector/quick-start/
- https://opentelemetry.io/docs/collector/installation/

References: 4755, 4396

## How has this been tested?

Ran locally and checked for the metrics in Honeycomb.

## Things to pay attention to during code review

Please describe parts of the change that require extra attention during code review, for example:

- File FFFF, line LL: This refactoring does this and this. Is it consistent with how it’s implemented elsewhere?
- Etc.

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [ ] I have added regression tests, if the PR fixes a bug
- [ ] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [ ] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

### CODE REVIEW

**Comment by vasconsaurus** [comment]
@caiosba I keep getting this error, I think I might have missed something silly, but I have no idea what

```
otel-collector-1     | 2024-06-18T19:47:41.088Z	debug	scrape/scrape.go:1331	Scrape failed	{"kind": "receiver", "name": "prometheus", "data_type": "metrics", "scrape_pool": "prometheus", "target": "http://0.0.0.0:3200/metrics", "error": "Get \"http://0.0.0.0:3200/metrics\": dial tcp 0.0.0.0:3200: connect: connection refused"}
```

What do you think?

**Comment by caiosba** [comment]
> @caiosba I keep getting this error, I think I might have missed something silly, but I have no idea what
> 
> ```
> otel-collector-1     | 2024-06-18T19:47:41.088Z	debug	scrape/scrape.go:1331	Scrape failed	{"kind": "receiver", "name": "prometheus", "data_type": "metrics", "scrape_pool": "prometheus", "target": "http://0.0.0.0:3200/metrics", "error": "Get \"http://0.0.0.0:3200/metrics\": dial tcp 0.0.0.0:3200: connect: connection refused"}
> ```
> 
> What do you think?

Is there any place you configure a host or something? I would expect it to be localhost:3200 or 127.0.0.1:3200

**Comment by vasconsaurus** [comment]
> > @caiosba I keep getting this error, I think I might have missed something silly, but I have no idea what
> > ```
> > otel-collector-1     | 2024-06-18T19:47:41.088Z	debug	scrape/scrape.go:1331	Scrape failed	{"kind": "receiver", "name": "prometheus", "data_type": "metrics", "scrape_pool": "prometheus", "target": "http://0.0.0.0:3200/metrics", "error": "Get \"http://0.0.0.0:3200/metrics\": dial tcp 0.0.0.0:3200: connect: connection refused"}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What do you think?
> 
> Is there any place you configure a host or something? I would expect it to be localhost:3200 or 127.0.0.1:3200

Yes. In the otel-collector-config.yaml. Here (example file): https://github.com/meedan/pender/blob/7df2d9585cce340d261b03936a16a05c3c4a18db/otel-collector-config.example.yaml#L8

I followed this guide https://docs.honeycomb.io/integrations/metrics/prometheus/, so I added `targets: ["0.0.0.0:3200"]`. I tested with localhost and 127.0.0.1, but I still get the same error.

**Comment by caiosba** [comment]
Prometheus and Pender are running in the same Docker container, right?

If so, are you able to manually reach the endpoint from inside the Pender container?

`curl -v http://localhost:3200/metrics`

**Comment by vasconsaurus** [comment]
> Prometheus and Pender are running in the same Docker container, right?
> 
> If so, are you able to manually reach the endpoint from inside the Pender container?
> 
> curl -v http://localhost:3200/metrics
> 

@caiosba with localhost it seems to be refused at first but with 0.0.0.0:3200 it goes straight through.

```
root@2afb72eb8a56:/app# curl -v http://localhost:3200/metrics
*   Trying ::1:3200...
* connect to ::1 port 3200 failed: Connection refused
*   Trying 127.0.0.1:3200...
* Connected to localhost (127.0.0.1) port 3200 (#0)
> GET /metrics HTTP/1.1
> Host: localhost:3200
> User-Agent: curl/7.74.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< content-type: text/plain; version=0.0.4
< Vary: Accept-Encoding
< Content-Length: 2994
<
# TYPE media_request_total counter
# HELP media_request_total Count every request made
# TYPE http_server_requests_total counter
# HELP http_server_requests_total The total number of HTTP requests handled by the Rack application.
http_server_requests_total{code="200",method="get",path="/"} 1.0
http_server_requests_total{code="200",method="get",path="/metrics"} 1.0
# TYPE http_server_request_duration_seconds histogram
# HELP http_server_request_duration_seconds The HTTP response duration of the Rack application.
http_server_request_duration_seconds_bucket{method="get",path="/",le="0.005"} 0.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="0.01"} 0.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="0.025"} 0.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="0.05"} 0.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="0.1"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="0.25"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="0.5"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="1"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="2.5"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="5"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="10"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/",le="+Inf"} 1.0
http_server_request_duration_seconds_sum{method="get",path="/"} 0.07536982700003136
http_server_request_duration_seconds_count{method="get",path="/"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="0.005"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="0.01"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="0.025"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="0.05"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="0.1"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="0.25"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="0.5"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="1"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="2.5"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="5"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="10"} 1.0
http_server_request_duration_seconds_bucket{method="get",path="/metrics",le="+Inf"} 1.0
http_server_request_duration_seconds_sum{method="get",path="/metrics"} 0.0009073799992620479
http_server_request_duration_seconds_count{method="get",path="/metrics"} 1.0
# TYPE http_server_exceptions_total counter
# HELP http_server_exceptions_total The total number of exceptions raised by the Rack application.
* Connection #0 to host localhost left intact
```

```
root@2afb72eb8a56:/app# curl -v http://0.0.0.0:3200/metrics
*   Trying 0.0.0.0:3200...
* Connected to 0.0.0.0 (127.0.0.1) port 3200 (#0)
> GET /metrics HTTP/1.1
> Host: 0.0.0.0:3200
> User-Agent: curl/7.74.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< content-type: text/plain; version=0.0.4
< Vary: Accept-Encoding
< Content-Length: 2994
<
# TYPE media_request_total counter
# HELP media_request_total Count every request made
# TYPE http_server_requests_total counter
# HELP http_server_requests_total The total number of HTTP requests handled by the Rack application.
http_server_requests_total{code="200",method="get",path="/"} 1.0
http_server_requests_total{code="200",method="get",path="/metrics"} 3.0
...
```

**Comment by caiosba** [comment]
Yeah, so 0.0.0.0, 127.0.0.1, localhost... all these should work. Maybe Prometheus is starting before Pender is up?

**Comment by vasconsaurus** [comment]
> Yeah, so 0.0.0.0, 127.0.0.1, localhost... all these should work. Maybe Prometheus is starting before Pender is up?

I have added `depends on: pender` to the otel-collector, which should keep the collector from starting first, right?
It polls every 15 seconds, even when I can reach the metrics endpoint it still returns connection refused.

**Comment by caiosba** [comment]
Aaaahhhh, I see now that `otel-collector` is running from a different container... OK, so in that case it needs to connect to `pender:3200`, not `localhost:3200`.

**Comment by caiosba** [comment]
(I thought it was a process that executed side-by-side with Pender at the Rack level)

If the only way is as a separate service, we'll need to check how to deploy this to production.

**Comment by vasconsaurus** [comment]
> Aaaahhhh, I see now that otel-collector is running from a different container... OK, so in that case it needs to connect to pender:3200, not localhost:3200.

I knew I was missing something simple 😅 
Seems to be working and talking to Honeycomb.

Screenshot 2024-06-19 at 11 14 02

⬆️ from the Honeycomb Interface


> (I thought it was a process that executed side-by-side with Pender at the Rack level)
> If the only way is as a separate service, we'll need to check how to deploy this to production.

So, there are two parts to this. The process that gets executed side-by-side with Pender at the Rack level is the Prometheus client, the one from this PR (https://github.com/meedan/pender/pull/455).
In order to send those metrics to Honeycomb we need the otel-collector, which is this one, and yes, a separate service. It scrapes the Prometheus endpoint and sends it.

**Comment by caiosba** [comment]
Got it. OK, great work Manu! When we have it all running locally correctly, let's talk to DevOps to see how this can get deployed.

File: config/otel-collector-config.example.yaml
Comment (caiosba) [inline config/otel-collector-config.example.yaml:1]: For consistency sake, please rename this file to `config/otel-collector-config.yaml.example`

**Review by caiosba — CHANGES_REQUESTED** [review]
Thanks Manu, just one minor change request. And one more question: Does it crash if the config file is not present?

**Comment by vasconsaurus** [comment]
> Thanks Manu, just one minor change request. And one more question: Does it crash if the config file is not present?

Yes. If I try to run `docker compose up` without the config it errors:

```
Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting "/.../pender/config/otel-collector-config.yaml" to rootfs at "/etc/otelcol-contrib/config.yaml": mount /.../pender/config/otel-collector-config.yaml:/etc/otelcol-contrib/config.yaml (via /proc/self/fd/6), flags: 0x5000: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type
```

**Comment by caiosba** [comment]
OK, sounds good. It's OK to leave it as is for this repo, but for the `check` repo, I don't think every developer needs this container to be running all the time, so my suggestion is to leave it commented in the `docker-compose.yml` file there... but that's for another PR :)

**Review by caiosba — APPROVED** [review]

**Comment by vasconsaurus** [comment]
Tests: We had one failure, a Facebook integration/contract test. 
I ran the test locally and it worked, talked to @caiosba. This looks like it was an intermittent error.
So I'm going ahead and merging it.

**Pull request merged**